nixos/kubernetes: allow configuring cfssl API server SANs

anmonteiro opened the pull request:
#### Motivation for this change

`services.kubernetes.easyCerts` is really useful to bootstrap a Kubernetes cluster. A common use case in Kubernetes clusters is to have a DNS entry for the master so that nodes can connect to a master that may change address (e.g. in an autoscaling group).

Right now this is not easily possible because the master also runs the cfssl API server but doesn't allow registering Subject Alternative Names in its TLS cert. 

This PR proposes allowing configuring the extra SANs in this certificate similarly to what's available in the API server:
 https://github.com/NixOS/nixpkgs/blob/a3465d9d104149c810ece20bb706395321fa8631/nixos/modules/services/cluster/kubernetes/apiserver.nix#L139-L143

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @johanot @calbrecht (apologies if spammy, this is my first contribution to nixpkgs and you both were listed as the top committers in `pki.nix`)

anmonteiro committed:
nixos/kubernetes: allow configuring cfssl API server SANs

anmonteiro commented:
Friendly ping @calbrecht @globin

anmonteiro commented:
Hey @jonringer, pinging you on this one for a review, feel free to ignore if it's too out of your depth.

jonringer commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -20,6 +20,7 @@ let
         size = 2048;
     };
     CN = top.masterAddress;
+    hosts = cfg.cfsslAPIExtraSANs;
```
would be nice to have a tests which demonstrates that this gets correctly used

jonringer reviewed (commented):
you're right, mostly out of my league

jonringer commented:
also, i think the commit/PR name should be:
```
nixos/kubernetes: Allow configuring cfssl API server SANs 
```

Infinisil commented on nixos/modules/services/cluster/kubernetes/pki.nix:
```diff
@@ -66,6 +67,14 @@ in
+      description = ''
+        Extra x509 Subject Alternative Names to be added to the cfssl API webserver TLS cert.
+      '';
+      default = [];
```
An example would be nice

anmonteiro commented:
Thanks for the initial review. I did the following things:

- rebased on current master
- changed the commit message as per @jonringer's review
- added an example as per @Infinisil 's suggestion

I can look into adding a test next, but it might take me a while to do that.

jonringer commented:
please squash commits

anmonteiro commented:
Done

jonringer reviewed (approved):
this seems to correspond with official documentation https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#create-a-certificate-signing-request

LGTM

jonringer commented:
@GrahamcOfBorg test kubernetes.dns
@GrahamcOfBorg test kubernetes.rbac
